File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rresultptr.py
#  @author Vincenzo Eduardo Padulano
#  @author Enric Tejedor
#  @date 2021-10


################################################################################
# Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.                      #
# All rights reserved.                                                         #
#                                                                              #
# For the licensing terms see $ROOTSYS/LICENSE.                                #
# For the list of contributors see $ROOTSYS/README/CREDITS.                    #
################################################################################

from ROOT import pythonization

@pythonization()
def pythonize_rresultptr(klass, name):
    """
    Pythonization for RResultPtr that always releases the GIL when triggering
    the RDF computation graph.
    """

    if name.startswith("ROOT::RDF::RResultPtr<"):
        klass.GetValue.__release_gil__ = True

    return True

File: tree/dataframe/src/RResultPtr.cxx
// Author: Vincenzo Eduardo Padulano, Enrico Guiraud CERN 2023/02

/*************************************************************************
 * Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.               *
 * All rights reserved.                                                  *
 *                                                                       *
 * For the licensing terms see $ROOTSYS/LICENSE.                         *
 * For the list of contributors see $ROOTSYS/README/CREDITS.             *
 *************************************************************************/

#include <string>

#include <ROOT/RResultPtr.hxx>
#include <ROOT/RDF/RInterface.hxx>
#include <ROOT/RDF/RLoopManager.hxx>

ROOT::Internal::RDF::SnapshotPtr_t
ROOT::Internal::RDF::CloneResultAndAction(const ROOT::Internal::RDF::SnapshotPtr_t &inptr,
                                          const std::string &outputFileName)
{
   return ROOT::Internal::RDF::SnapshotPtr_t{
      inptr.fObjPtr, inptr.fLoopManager,
      inptr.fActionPtr->CloneAction(reinterpret_cast<void *>(const_cast<std::string *>(&outputFileName)))};
}
